ext/drmgr/drmgr.h

```diff
@@ -947,25 +947,109 @@ drmgr_get_emulated_instr_data(instr_t *instr, OUT emulated_instr_t *emulated);
+ *         record_instr_fetch(instr_fetch);
+ *     instr_t *instr_operands = drmgr_orig_app_instr_for_operands(drcontext);
+ *     if (instr_operands_valid != NULL)
+ *         record_data_addresses(instr_operands);
+ *     return DR_EMIT_DEFAULT;
+ * }
+ *
+ * \return false if the caller's \p emulated_instr_t is not compatible, true otherwise.
+ */
+DR_EXPORT
+bool
+drmgr_in_emulation_region(void *drcontext, OUT const emulated_instr_t **emulation_info);
+
+DR_EXPORT
+/**
+ * Must be called during drmgr's insertion phase.
+ *
+ * Returns the instruction to consider as the current application
+ * instruction for observational clients with respect to which
+ * instruction executed, taking into account emulation.  This may be
+ * different from the current instruction list instruction passed to
+ * the insertion event.
+ *
+ * Returns NULL if observational clients should skip the current instruction
+ * list instruction.
+ *
+ * This is a convenience routine.  It follows the logic of the recommended usage
+ * for drmgr_in_emulation_region() laid out in the documentation for that method,
+ * repeated here for instruction fetch only:
+ *
  *    const emulated_instr_t *emulation;
  *    if (drmgr_in_emulation_region(drcontext, &emulation)) {
  *        if (TEST(DR_EMULATE_FIRST_INSTR, emulation->flags)) {
- *            record_instr_fetch(emulation->instr);
- *            if (!TEST(DR_EMULATE_INSTR_ONLY, emulation->flags))
- *                record_data_addresses(emulation->instr);
+ *            return emulation->instr;
  *        } // Else skip further instr fetches until outside emulation region.
- *        if (instr_is_app(exec_instr) && TEST(DR_EMULATE_INSTR_ONLY, emulation->flags))
- *            record_data_addresses(exec_instr);
- *    } else if (instr_is_app(exec_instr)) {
- *        record_instr_fetch(exec_instr);
- *        record_data_addresses(exec_instr);
+ *    } else if (instr_is_app(insert_instr)) {
+ *        return insert_instr;
  *    }
- *
- * \return false if the caller's \p emulated_instr_t is not compatible, true otherwise.
+ *    return NULL;
  */
+instr_t *
+drmgr_orig_app_instr_for_fetch(void *drcontext);
+
 DR_EXPORT
-bool
-drmgr_in_emulation_region(void *drcontext, OUT const emulated_instr_t **emulation_info);
+/**
+ * Must be called during drmgr's insertion phase.
+ *
+ * Returns the instruction to consider as the current application
+ * instruction for observational clients with respect to the operands
+ * of the instruction, taking into account emulation.  This may be
+ * different from the current instruction list instruction passed to
+ * the insertion event.
+ *
+ * Returns NULL if observational clients should skip the current instruction
+ * list instruction.
+ *
+ * This is a convenience routine.  It follows the logic of the recommended usage
+ * for drmgr_in_emulation_region() laid out in the documentation for that method,
+ * repeated here for operands only:
+ *
+ *    const emulated_instr_t *emulation;
```

Comment:
Maybe remove the code inlined in the documentation here? The reader can always look at its implementation.

Reply:
Done.